File: common/src/main/java/com/google/tsunami/common/net/http/HttpResponse.java
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.tsunami.common.net.http;

import com.google.auto.value.AutoValue;
import com.google.auto.value.extension.memoized.Memoized;
import com.google.errorprone.annotations.Immutable;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.google.protobuf.ByteString;
import java.util.Optional;
import okhttp3.HttpUrl;

/** Immutable HTTP response. */
@Immutable
@AutoValue
@AutoValue.CopyAnnotations
// HttpUrl is immutable even if not marked as such.
@SuppressWarnings("Immutable")
public abstract class HttpResponse {
  public abstract HttpStatus status();
  public abstract HttpHeaders headers();
  public abstract Optional<ByteString> bodyBytes();
  // The URL that produced this response.
  // TODO(b/173574468): Provide the full redirection request not just the Url.
  public abstract Optional<HttpUrl> responseUrl();

  /**
   * Gets the body of the HTTP response as a UTF-8 encoded String.
   *
   * @return HTTP response body as a Java {@link String}.
   */
  @Memoized
  public Optional<String> bodyString() {
    return bodyBytes().map(ByteString::toStringUtf8);
  }

  /**
   * Tries to parse the response body as json and returns the parsing result as {@link JsonElement}.
   * If parsing failed, an empty optional is returned.
   *
   * @return HTTP response body as a Gson {@link JsonElement} object.
   */
  @Memoized
  public Optional<JsonElement> bodyJson() {
    try {
      return bodyString().map(JsonParser::parseString);
    } catch (RuntimeException e) {
      // Do best-effort parsing and ignore Json parsing errors
      return Optional.empty();
    }
  }

  /**
   * Tries to determine if a given field in Json response is equal to a specific value. If parsing
   * failed, {@link com.google.gson.JsonSyntaxException} or {@link IllegalStateException} will be
   * thrown.
   *
   * @return boolean
   */
  public boolean jsonFieldEqualsToValue(String fieldname, String value) {
    Optional<JsonPrimitive> jsonPrimitive =
        bodyJson()
            .map(JsonElement::getAsJsonObject)
            .map(object -> object.getAsJsonPrimitive(fieldname));
    return jsonPrimitive.isPresent() && jsonPrimitive.get().getAsString().equals(value);
  }

  public static Builder builder() {
    return new AutoValue_HttpResponse.Builder();
  }

  /** Builder for {@link HttpResponse}. */
  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setStatus(HttpStatus httpStatus);
    public abstract Builder setHeaders(HttpHeaders httpHeaders);
    public abstract Builder setBodyBytes(ByteString bodyBytes);
    public abstract Builder setBodyBytes(Optional<ByteString> bodyBytes);
    public abstract Builder setResponseUrl(HttpUrl url);
    public abstract Builder setResponseUrl(Optional<HttpUrl> url);

    public abstract HttpResponse build();
  }
}

File: plugin_server/py/common/net/http/http_response.py
"""HTTP response utility."""

import abc
import json
from typing import Any, Optional
from tsunami.plugin_server.py.common.net.http.http_headers import HttpHeaders
from tsunami.plugin_server.py.common.net.http.http_status import HttpStatus


class HttpResponse(metaclass=abc.ABCMeta):
  """HTTP request utility class.

  Please use Builder() to create instances of this class.

  Attributes:
    status: The HTTP response status.
    url: String address that produced this HTTP response.
    headers: The HTTP response headers in key/value pairs.
    body: The HTTP body data.
  """

  def __init__(self):
    self.status: HttpStatus = None
    self.url: Optional[str] = None
    self.headers: HttpHeaders = None
    self.body: Optional[bytes] = None

  def body_string(self) -> str:
    """Get the body data as a UTF-8 encoded string."""
    try:
      return self.body.decode('utf-8')
    except AttributeError:
      return ''

  def body_json(self) -> Optional[dict[str, Any]]:
    """Parse the response body as JSON. Returns null if parsing failed."""
    try:
      return json.loads(self.body_string())
    except ValueError:
      return None

  def json_field_has_value(self, field: str, value: str):
    """Check if JSON body of the response has field and value pair."""
    body = self.body_json()
    if body is not None:
      return body.get(field) == value
    return False

  @classmethod
  def builder(cls):
    return Builder()


class Builder:
  """Builder class to create HTTP request object.

  Attributes:
    http_response: HTTP request object built.
  """

  def __init__(self):
    self.http_response = HttpResponse()

  def set_status(self, status: HttpStatus):
    """Set the HttpStatus type."""
    self.http_response.status = status
    return self

  def set_headers(self, headers: HttpHeaders):
    """Set the HttpHeaders for the response."""
    self.http_response.headers = headers
    return self

  def set_response_body(self, response_body: Optional[bytes]):
    """Set the body for the response."""
    self.http_response.body = response_body
    return self

  def set_url(self, url: str):
    """Set URL response."""
    self.http_response.url = url
    return self

  def build(self):
    return self.http_response
